src/internal/_equals.js

```diff
@@ -27,6 +28,10 @@ module.exports = function _equals(a, b, stackA, stackB) {
     case 'Arguments':
     case 'Array':
     case 'Object':
+      if (typeof a.constructor === 'function' &&
+          _functionName(a.constructor) === 'Promise') {
```

Comment:
Unfortunately `Object.prototype.toString.call(Promise.resolve(42))` evaluates to `'[object Object]'` rather than `'[object Promise]'`, so we're forced to do something nasty.